### PR TITLE
Retain operation level parameters over path level ones

### PR DESCRIPTION
**Pull request by arthurcro**
### Motivation

Fixes https://github.com/apple/swift-openapi-generator/issues/168.

### Modifications

Previously, an operation's parameters were returned as a concatenation of the parameters from the path item level with those at operation level.

However, this is incorrect according to https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#fixed-fields-7. 
> A list of parameters that are applicable for all the operations described under this path. These parameters can be overridden at the operation level, but cannot be removed there. The list MUST NOT include duplicated parameters.

With this change, we are merging the two arrays of parameters using a unique identifier of the location + name of parameters. If duplicate parameters exist, only the parameters from the operation level are preserved.

### Result

After this change, an operation's parameters won't contain any duplicate. 

### Test Plan

Add tests for `OperationDescription.allParameters` at `Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift`.

### CODE REVIEW

File: Sources/_OpenAPIGeneratorCore/Translator/Operations/OperationDescription.swift [hunk 1]
```diff
@@ -144,10 +144,29 @@ extension OperationDescription {
...
-    /// and the operation level.
+    /// Merged parameters from both the path item level and the operation level.
+    /// If duplicate parameters exist, only the parameters from the operation level are preserved.
+    ///
+    /// - Returns: An array of merged path item and operation level parameters without duplicates.
+    /// - Throws: When an invalid JSON reference is found.
     var allParameters: [UnresolvedParameter] {
-        pathParameters + operation.parameters
+        get throws { 
+            var mergedParameters = [UnresolvedParameter]()
```
Comment (czechboy0) [inline Sources/_OpenAPIGeneratorCore/Translator/Operations/OperationDescription.swift:154]:
```suggestion
            var mergedParameters: [UnresolvedParameter] = []
```

Nit: prefer a literal.

File: Sources/_OpenAPIGeneratorCore/Translator/Operations/OperationDescription.swift [hunk 2]
```diff
@@ -144,10 +144,29 @@ extension OperationDescription {
...
+            var uniqueIdentifiers = Set<String>()
```
Comment (czechboy0) [inline Sources/_OpenAPIGeneratorCore/Translator/Operations/OperationDescription.swift:155]:
```suggestion
            var mergedParameters: [UnresolvedParameter] = []
            var uniqueIdentifiers: Set<String> = []
```

Style nit: prefer a literal.

File: Sources/_OpenAPIGeneratorCore/Translator/Operations/OperationDescription.swift [hunk 3]
```diff
@@ -144,10 +144,29 @@ extension OperationDescription {
...
+
+            for parameter in pathParameters + operation.parameters {
+                let resolvedParameter = try parameter.resolve(in: components)
+                let identifier = resolvedParameter.name + resolvedParameter.location.rawValue
```
Comment (czechboy0) [inline Sources/_OpenAPIGeneratorCore/Translator/Operations/OperationDescription.swift:159]:
```suggestion
                let identifier = resolvedParameter.location.rawValue + ":" + resolvedParameter.name
```

Nit: it reads better if the larger scope (location) goes first, then the name that only needs to be unique within the location, and let's separate it with a character for easier readability when debugging.

File: Sources/_OpenAPIGeneratorCore/Translator/Operations/OperationDescription.swift [hunk 4]
```diff
@@ -144,10 +144,29 @@ extension OperationDescription {
...
+
+                guard !uniqueIdentifiers.contains(identifier) 
+                else { continue }
```
Comment (czechboy0) [inline Sources/_OpenAPIGeneratorCore/Translator/Operations/OperationDescription.swift:162]:
```suggestion
                guard !uniqueIdentifiers.contains(identifier) else { 
                    continue 
                }
```

Style nit.

**Comment by czechboy0** [comment]
The change looks good, just a few minor style suggestions.

For testing, you can create a new unit test file `Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift`, follow the style seen e.g. in `Tests/OpenAPIGeneratorCoreTests/Translator/Content/Test_ContentType.swift`, and create an `OperationDescription` with duplicated params in the path and operation in the unit test and verify that `allParameters` returns just the unique ones.

**Comment by arthurcro** [comment]
Thank you for the nice review, I applied all suggestions! 🙂
I'll add a test shortly and mark this as ready for review.

File: Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import OpenAPIKit30
+import XCTest
+@testable import _OpenAPIGeneratorCore
+
+final class Test_OperationDescription: Test_Core {
+
+    func testAllParameters_duplicates_retainOnlyOperationParameters() throws {
```
Comment (arthurcro) [inline Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:20]: @czechboy0 I'm not sure about usign the `YamsParser` here. I could not find an easier way to construct an `OperationDescription`. It needs an `OpenAPI.Endpoint` but it does not have a `public` initializer. 
I thought about constructing a `OpenAPI.Document` or `OpenAPI.PathItem`manually but I found it make the test harder to read and understand. What do you think? 🙂

Comment (czechboy0) [reply Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:20]: I think constructing `OpenAPI.PathItem` is better than creating a document from YAML. See if you can get that to work to initialize an `OperationDescription`.

Comment (czechboy0) [reply Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:20]: From `OpenAPI.PathItem`, you can get the `OpenAPI.PathItem.Endpoint` for your tested operation.

**Comment by czechboy0** [comment]
@swift-server-bot add to allowlist

File: Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift (hunk 5 above)
Comment (arthurcro) [reply Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:20]: Alright, thanks for the feedback. I'll work on updating the tests to use that instead!

Comment (arthurcro) [reply Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:20]: I've pushed the changes. Thanks to the tests I also noticed I was retaining the path parameter over the operation ones which was incorrect according to https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#path-item-object so I fixed that too!

Comment (czechboy0) [reply Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:20]: I think the order was right before, the shared path params should come before the operation specific ones.

If you flip the order back, you should be able to just slightly change your logic - do (pathParams + operationParams), get a reversed(), iterate that, then reverse the result again. I think that should get us the original order but also retain the last occurrence of any location+name pair.

Comment (arthurcro) [reply Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:20]: Ah right, the order should be kept as well. I'll adjust that part, thank you!

File: Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    private func _test(_ pathItem: OpenAPI.PathItem) throws -> [UnresolvedParameter] {
+        guard let endpoint = pathItem.endpoints.first else {
+            XCTFail("Unable to retrieve the path item first endpoint.")
+            return []
+        }
+
+        let operationDescription = OperationDescription(
+            path: .init(["\test"]),
```
Comment (czechboy0) [inline Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:87]:
```suggestion
            path: .init(["/test"]),
```

File: Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                parameters: [operationLevelParameter],
+                requestBody: .b(.init(content: [:])),
+                responses: [:]
+            ),
+            vendorExtensions: [:]
+        )
+        let allParameters = try _test(pathItem)
+
+        XCTAssertEqual(allParameters, [operationLevelParameter])
+    }
```
Comment (czechboy0) [inline Tests/OpenAPIGeneratorCoreTests/Translator/Operations/Test_OperationDescription.swift:48]: Can you please add one more test that verifies the correct ordering?

Could be:
- path item level: A, B
- operation level: B, C
- expected result: A, B, C

**Review by czechboy0 — CHANGES_REQUESTED** [review]
Just two more comments about the tests, otherwise looks great!

**Pull request merged**